### DISCUSSION TITLE
How to Generate a PDF

### DISCUSSION THREAD

**Discussion by R2bEEaton** [Budibase/budibase#7278] [Show and tell]
This is unbelievably hacky. Hopefully we get official support soon. **HUGE downside is you can't select any text in the PDF. You effectively have taken a screenshot of a part of the page downloaded as a PDF.**

Here is a brief explanation of how this works. Comment or check the included export if you have any questions.
1. [html2pdf](https://github.com/eKoopmans/html2pdf.js) is a JavaScript function that lets you convert an HTML element to a PDF.
2. To get this into Budibase, download [`dist/html2pdf.bundle.min.js`](https://github.com/eKoopmans/html2pdf.js/blob/master/dist/html2pdf.bundle.min.js) and upload it to a table in Budibase with an attachment field. Then get the URL of the file by right clicking on the preview on the table and copying the link. Uploading the script to a Budibase table is to get around an untrusted script source problem later on.
3. Import this script using an Embed element, making use of Budibase XSS as described in #6839.
4. Create and name a container to convert to a PDF. This will contain both the content and the look of your PDF.
5. Make a download PDF button using an Embed component. You can style this however you want and make it look like a Budibase button too if you want. Be sure to change `container1` to whatever you named your container. All of this code below is necessary to include Budibase CSS when only converting a single container to PDF. If you want to do the whole page, it would be as simple as doing `onclick="html2pdf(document);"`.
```
<button onclick="
var elementToPrint = document.createElement('html');
elementToPrint.appendChild(document.head.cloneNode(true));
var spectrumRoot = document.createElement('div');
spectrumRoot.className = 'spectrum spectrum--medium spectrum--light svelte-rrn5ov';
spectrumRoot.appendChild(document.querySelector(&quot;div[data-name='container1']&quot;).cloneNode(true));
elementToPrint.appendChild(spectrumRoot);
html2pdf(elementToPrint);
">Generate PDF of container1</button>
```

The PDF generation will take 5-10 seconds and even longer for the whole document, but it should start downloading automatically. You should probably display some waiting animation while it creates the PDF.

Here is the export. Please consult this if you have any confusion on how to implement the steps above.
[PDF Hack-export-1660576133844.txt](https://github.com/Budibase/budibase/files/9338080/PDF.Hack-export-1660576133844.txt)

**The export was made on v1.2.20 and tested on Microsoft Edge.**

**Comment by kasgithub8** [comment]
Hi @R2bEEaton,

the Import doesn't work for me❓

**Reply by R2bEEaton** [reply]
In what way? Also the export was made using v1.2.20.

**Reply by kasgithub8** [reply]
> In what way? 

Well, I exported and imported it and the button didn't work.

> Also the export was made using v1.2.20.

That's probably why.

**Reply by R2bEEaton** [reply]
Ok. The button should take 5-10 seconds to result in anything, so make sure you're giving it enough time. Also check the JavaScript console using your browser for any errors and let me know what you see. Which browser are you using?

**Reply by kasgithub8** [reply]
Chrome & 
![image](https://user-images.githubusercontent.com/95284278/184884854-c6693e9c-58a0-4426-b544-709a12e1dea1.png)

**Reply by R2bEEaton** [reply]
Okay, you might need to upload and download the html2pdf file to the
database yourself and re-link it yourself by modifying the html2pdf embed
in the component tree.

>[html2pdf](https://github.com/eKoopmans/html2pdf.js) is a JavaScript function that lets you convert an HTML element to a PDF.
When I follow your link, do I have to download a certain file? If not, which one do I upload?

The dist one from step 2.

(sorry I accidentally made new comments instead of replying)

**Reply by kasgithub8** [reply]
Ooh, got it now. This is neat. Thank u for explaining!

**Comment by R2bEEaton** [comment]
Okay, you might need to upload and download the html2pdf file to the
database yourself and re-link it yourself by modifying the html2pdf embed
in the component tree.

[See @melohagan's comment.](https://github.com/Budibase/budibase/discussions/7278#discussioncomment-3406928)

**Reply by kasgithub8** [reply]
> 1. [html2pdf](https://github.com/eKoopmans/html2pdf.js) is a JavaScript function that lets you convert an HTML element to a PDF.

When I follow your link, do I have to download a certain file? If not, which one do I upload?

**Comment by melohagan** [comment]
This is very nice - great job! 👍 

File attachments are stored on your own MinIO instance, therefore to get this to work you will need to replace the file attachment. 

### Step 1 - Upload html2pdf
Open this link in your browser: https://cdnjs.cloudflare.com/ajax/libs/html2pdf.js/0.10.1/html2pdf.bundle.min.js
Press Ctrl+S to download the file.

Edit the *jsfile* table row. Remove the existing attachment and upload the downloaded minified js.

### Step 2 - Update the import embed
Click on the JS file to open the attachment in a new browser tab.
![Screenshot 2022-08-16 at 15 00 54](https://user-images.githubusercontent.com/101575380/184898758-a5e8b841-0e26-4169-8cbf-7f3f1caa48f9.png)

Copy the URL from `/prod` onwards:
![Screenshot 2022-08-16 at 15 02 15](https://user-images.githubusercontent.com/101575380/184899068-9518cb2a-2146-484f-8464-9c9a4226d78c.png)

Replace the addScript URL in the import embed:
![Screenshot 2022-08-16 at 15 03 03](https://user-images.githubusercontent.com/101575380/184899234-22f8a363-d4eb-43b7-a2b1-5b525149be14.png)

### Result
![pdf](https://user-images.githubusercontent.com/101575380/184899538-ddc4ce1d-7a86-47cf-9512-3ac229cf0e4e.gif)

**Reply by R2bEEaton** [reply]
Good catch, thanks for this!

**Reply by rafikApp** [reply]
Hey @R2bEEaton 
Can we generate the pdf using an automation.
For example when we click on (save). we will save the row and generate the pdf

**Comment by roafhtun** [comment]
Is it still working? I cannot get it to work thats why.

Thanks!
